File: sally/leetcode/98_Validate Binary Search Tree.cpp
// 98. Validate Binary Search Tree
// Runtime: 8 ms, faster than 90.29% of C++ online submissions for Validate Binary Search Tree.
// Memory Usage: 21.6 MB, less than 69.86% of C++ online submissions for Validate Binary Search Tree.


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        stack<pair<TreeNode*, pair<long long int, long long int>>> st;
        TreeNode* cur;
        st.push({root, {-pow(2, 31)-1, pow(2, 31)}});
        long long int max_num, min_num;
        
        while(!st.empty()){
            cur = st.top().first;
            min_num = st.top().second.first;
            max_num = st.top().second.second;
            
            if(cur->val <= min_num || cur->val >= max_num) return false;
            
            st.pop();
            if(cur->left != nullptr){
                if(cur->left->val >= cur->val) return false;
                st.push({cur->left, {min_num, cur->val}});
            }
            if(cur->right != nullptr){
                if(cur->right->val <= cur->val) return false;
                st.push({cur->right, {cur->val, max_num}});  
            }
        }
        return true;
    }
};

File: opijae/leetcode/98_Validate_Binary_Search_Tree.py
from typing import *

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.validBST(root, float('-inf'), float('inf'))
    def validBST(self, node, _min, _max):
        """
        Args:
            node : 현재 root node
            _min : 임계값(작은)
            _max : 임계값(큰)
        """
        if not node:
            return True
        
        if node.val <= _min or node.val >= _max:
            return False
        return self.validBST(node.left, _min, node.val) and self.validBST(node.right, node.val, _max)